app/controllers/PreviouslyRegisteredController.scala

```diff
@@ -49,27 +52,61 @@ class PreviouslyRegisteredController @Inject()(
       Ok(view(form, AdministratorOrPractitioner.Practitioner))
   }
 
-  def onSubmitAdministrator: Action[AnyContent] = Action {
+  def onSubmitAdministrator: Action[AnyContent] = Action.async {
     implicit request =>
       form.bindFromRequest().fold(
         (formWithErrors: Form[_]) =>
-          BadRequest(view(formWithErrors, AdministratorOrPractitioner.Administrator)),
+          Future.successful(BadRequest(view(formWithErrors, AdministratorOrPractitioner.Administrator))),
         {
-          case PreviouslyRegisteredButNotLoggedIn => Redirect(appConfig.recoverCredentialsPSAUrl)
-          case _ => Redirect(appConfig.registerSchemeAdministratorUrl)
+          case PreviouslyRegisteredButNotLoggedIn =>
+            authorised().retrieve(Retrievals.allEnrolments) {
+              case enrolments =>
+                isTpssAccountPSA(enrolments) match {
+                  case true => Future.successful(Redirect(routes.TpssRecoveryController.onPageLoad))
+                  case false => Future.successful(Redirect(appConfig.recoverCredentialsPSAUrl))
+                }
+              case _ =>
+                Future.successful(Redirect(appConfig.recoverCredentialsPSAUrl))
+            }
+          case _ => Future.successful(Redirect(appConfig.registerSchemeAdministratorUrl))
         }
       )
   }
 
-  def onSubmitPractitioner: Action[AnyContent] = Action {
+  def onSubmitPractitioner: Action[AnyContent] = Action.async {
     implicit request =>
       form.bindFromRequest().fold(
         (formWithErrors: Form[_]) =>
-          BadRequest(view(formWithErrors, AdministratorOrPractitioner.Practitioner)),
+          Future.successful(BadRequest(view(formWithErrors, AdministratorOrPractitioner.Practitioner))),
         {
-          case PreviouslyRegisteredButNotLoggedIn => Redirect(appConfig.recoverCredentialsPSPUrl)
-          case _ => Redirect(appConfig.registerSchemePractitionerUrl)
+          case PreviouslyRegisteredButNotLoggedIn =>
```

Comment:
Related to above

Reply:
moved common code into separate method and re-used for PSAs and PSPs